src/main/java/org/broadinstitute/hellbender/tools/coveragemodel/CoverageModelSparkUtils.java

```diff
@@ -0,0 +1,151 @@
+package org.broadinstitute.hellbender.tools.coveragemodel;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.apache.commons.math3.util.FastMath;
+import org.apache.spark.api.java.JavaPairRDD;
+import org.broadinstitute.hellbender.utils.param.ParamUtils;
+import org.nd4j.linalg.api.ndarray.INDArray;
+import org.nd4j.linalg.factory.Nd4j;
+import org.nd4j.linalg.indexing.NDArrayIndex;
+import scala.Tuple2;
+
+import javax.annotation.Nonnull;
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Spark-related helper functions for {@link CoverageModelEMWorkspace}
+ *
+ * @author Mehrtash Babadi &lt;mehrtash@broadinstitute.org&gt;
+ */
+public final class CoverageModelSparkUtils {
+
+    /**
+     * Creates a uniform partitioning of an index space. If {@param minBlockSize} is larger than
+     * {@param length}/{@param numBlocks}, fewer partitions will be created.
+     *
+     * @param length the length of the index space
+     * @param numBlocks number of partitions (= blocks)
+     * @param minBlockSize minimum number of indices in each partition
+     * @return a list of {@link LinearSpaceBlock}
+     */
+    public static List<LinearSpaceBlock> createLinearSpaceBlocks(final int length, final int numBlocks,
+                                                                 final int minBlockSize) {
+        ParamUtils.isPositive(length, "The length of the linear space to be partitioned must be positive");
+        ParamUtils.isPositive(numBlocks, "The number of blocks must be positive");
+        ParamUtils.isPositive(minBlockSize, "Minimum block size must be positive");
+
+        final int blockSize = FastMath.max(length/numBlocks, minBlockSize);
+        final List<LinearSpaceBlock> blocks = new ArrayList<>();
+        for (int begIndex = 0; begIndex < length; begIndex += blockSize) {
+            blocks.add(new LinearSpaceBlock(begIndex, FastMath.min(begIndex + blockSize, length)));
+        }
+        /* the last block might be smaller than minBlockSize; we merge them */
+        while (blocks.size() > numBlocks && blocks.size() > 1) {
+            final int newBegIndex = blocks.get(blocks.size() - 2).getBegIndex();
+            final int newEndIndex = blocks.get(blocks.size() - 1).getEndIndex();
+            blocks.remove(blocks.size() - 1);
+            blocks.remove(blocks.size() - 1);
+            blocks.add(new LinearSpaceBlock(newBegIndex, newEndIndex));
+        }
+        return blocks;
+    }
+
+    /**
+     * Partitions an {@link INDArray} along the 0th axis according to a given partitioning and returns
+     * a list
+     *
+     * @param blocks partitioning
+     * @param arr an array
+     * @return list of partitioned arrays
+     */
+    public static List<Tuple2<LinearSpaceBlock, INDArray>> partitionINDArrayToAList(@Nonnull final List<LinearSpaceBlock> blocks,
```

Comment:
It is used in the regularizer part. Once I get to that problem again, I will revise the implementation. In principle, we can work with just maps.